### PR TITLE
azure_resource_info metric

### PR DESCRIPTION
**Pull request by rolandvarga**
This change has been opened in response to issue #64 

The PR introduces the following changes in order to create an azure_resource_info metric:
- dynamically identify appropriate Azure Monitor API version for each resource type (resource information for each type is exposed under a different endpoint, no version to support all resource types)
- retrieve resource infos for each collected resource (tags, resource type, etc...)
- expose resource infos as metric labels
- update prometheus GO client (and other dependencies) to allow different number of label dimensions in the same metric family

### CODE REVIEW

File: utils.go [hunk 1]
```diff
@@ -32,12 +43,40 @@ func GetTimes() (string, string) {
...
 
+	if len(split) > 13 {
+		labels["sub_resource_name"] = split[subResourceNamePosition]
+	}
+
+	// create a label for each field of the resource
+	val := reflect.ValueOf(rm.Resource)
+	for i := 0; i < val.NumField(); i++ {
+		field := val.Field(i)
+		tag := reflect.TypeOf(rm.Resource).Field(i).Tag.Get(formatTag)
```
Comment (brian-brazil) [inline utils.go]: Do these need sanitisation?

File: utils.go [hunk 2]
```diff
@@ -4,10 +4,21 @@ import (
...
 	"time"
 )
 
+var (
+	// resource component positions in a ResourceURL
+	resourceGroupPosition   = 4
+	resourceNamePosition    = 8
+	subResourceNamePosition = 10
+
+	invalidLabelChars = regexp.MustCompile("[^\\w]")
```
Comment (brian-brazil) [inline utils.go]: That's not quite the pattern, see https://github.com/prometheus/common/blob/master/model/metric.go#L27

File: utils.go [hunk 3]
```diff
@@ -32,12 +43,40 @@ func GetTimes() (string, string) {
...
+		k = strings.ToLower(k)
+		k = invalidLabelChars.ReplaceAllString(k, "_")
+		labels[k] = v
 	}
 
+	if len(split) > 13 {
+		labels["sub_resource_name"] = split[subResourceNamePosition]
+	}
+
+	// create a label for each field of the resource
```
Comment (brian-brazil) [inline utils.go]: What if there's a clash between this and the tags?

File: main.go [hunk 4]
```diff
@@ -42,23 +42,24 @@ func (c *Collector) Describe(ch chan<- *prometheus.Desc) {
 }
 
 type resourceMeta struct {
-	resourceURL  string
-	metrics      string
-	aggregations []string
+	ResourceURL  string
```
Comment (brian-brazil) [inline main.go]: Why make this public?

File: utils.go [hunk 5]
```diff
@@ -32,12 +43,40 @@ func GetTimes() (string, string) {
...
+	}
+
+	return labels
+}
+
+func CreateAllResourceLabelsFrom(rm resourceMeta) map[string]string {
+	formatTag := "pretty"
+	labels := make(map[string]string)
+	split := strings.Split(rm.ResourceURL, "/")
+	labels["resource_group"] = split[resourceGroupPosition]
```
Comment (brian-brazil) [inline utils.go]: What about the resource_name?

File: main.go [hunk 6]
```diff
@@ -69,40 +70,47 @@ func (c *Collector) extractMetrics(ch chan<- prometheus.Metric, resource resourc
...
 				metricValue.Maximum,
 			)
 		}
 	}
+
+	infoLabels := CreateAllResourceLabelsFrom(rm)
+	ch <- prometheus.MustNewConstMetric(
+		prometheus.NewDesc("azure_resource_info", "Azure information available for resource", nil, infoLabels),
+		prometheus.UntypedValue,
+		0,
```
Comment (brian-brazil) [inline main.go]: It should be a gauge with value 1

File: main.go [hunk 7]
```diff
@@ -243,6 +266,11 @@ func main() {
 		os.Exit(0)
 	}
 
+	err = ac.listAPIVersions()
```
Comment (brian-brazil) [inline main.go]: Why did you add this?

File: main.go [hunk 8]
```diff
@@ -196,13 +213,20 @@ func (c *Collector) Collect(ch chan<- prometheus.Metric) {
...
 			var resource resourceMeta
-			resource.metrics = metricsStr
-			resource.aggregations = filterAggregations(resourceTag.Aggregations)
-			resource.resourceURL = resourceURLFrom(f, resource.metrics, resource.aggregations)
+			resource.Metrics = metricsStr
+			resource.Aggregations = filterAggregations(resourceTag.Aggregations)
+			resource.ResourceURL = resourceURLFrom(f.ID, resource.Metrics, resource.Aggregations)
+
+			var err error
+			resource.Resource, err = ac.lookupResourceByID(f.ID)
```
Comment (brian-brazil) [inline main.go]: This will cause quite a few API calls. Can we get something without this?

File: utils.go (hunk 1 above)
Comment (rolandvarga) [reply utils.go]: The field tags are handled already in `AzureResource`, and the values are coming directly from Azure Monitor, so we think we should be good.

File: utils.go (hunk 2 above)
Comment (rolandvarga) [reply utils.go]: Did a correction based on your link

File: utils.go (hunk 3 above)
Comment (rolandvarga) [reply utils.go]: Good point. I added a prefix for tags.

File: main.go (hunk 4 above)
Comment (rolandvarga) [reply main.go]: Changed back to private

File: utils.go (hunk 5 above)
Comment (rolandvarga) [reply utils.go]: `resource_name` is a field in `AzureResource` and is added to the labels along with the objects other fields.

File: main.go (hunk 6 above)
Comment (rolandvarga) [reply main.go]: changed

File: main.go (hunk 7 above)
Comment (rolandvarga) [reply main.go]: The API version endpoint seems to be very different among resource types, hence we needed a way to match them. An initial approach was to configure a static list and get resource information based on the version numbers in that list. We thought it would be better to have a dynamically created list so that there's one less thing to maintain.

File: main.go (hunk 8 above)
Comment (rolandvarga) [reply main.go]: As an alternativa approach I introduced some caching so we query each resource only once. The logic itself is required to lookup the tags as only the resource group based config retrieves those by default.

**Comment by rolandvarga** [comment]
@brian-brazil made some changes based on your comments. Let me know what you think.

File: utils.go (hunk 5 above)
Comment (brian-brazil) [reply utils.go]: That needs a comment then.

File: main.go [hunk 9]
```diff
@@ -27,6 +27,8 @@ var (
 	invalidMetricChars    = regexp.MustCompile("[^a-zA-Z0-9_:]")
 	azureErrorDesc        = prometheus.NewDesc("azure_error", "Error collecting metrics", nil, nil)
 	batchSize             = 20
+	publishedResource     = map[string]bool{}
```
Comment (brian-brazil) [inline main.go]: These are globals without any locking, they should be per-scrape at most as I presume they can change over time.

File: main.go [hunk 10]
```diff
@@ -148,27 +153,33 @@ func (c *Collector) Collect(ch chan<- prometheus.Metric) {
...
-		if err != nil {
-			log.Printf("failed to get resource information for target %s: %v", target.Resource, err)
-			ch <- prometheus.NewInvalidMetric(azureErrorDesc, err)
-			continue
+		if _, ok := resourceList[target.Resource]; !ok {
+			resource, err := ac.lookupResourceByID(target.Resource)
+			if err != nil {
+				log.Printf("failed to get resource information for target %s: %v", target.Resource, err)
+				ch <- prometheus.NewInvalidMetric(azureErrorDesc, err)
+				continue
```
Comment (brian-brazil) [inline main.go]: You can return at this point, the scrape has failed.

File: main.go (hunk 9 above)
Comment (rolandvarga) [reply main.go]: ended up removing one of those globals, `publishedResource` is per-scrape now.

File: main.go (hunk 10 above)
Comment (rolandvarga) [reply main.go]: changed

**Comment by rolandvarga** [comment]
Hi @brian-brazil,
I made more changes based on your comments and also introduced batching in the newly added calls. Let me know what you think about them.

File: main.go [hunk 11]
```diff
@@ -44,13 +44,14 @@ func (c *Collector) Describe(ch chan<- *prometheus.Desc) {
...
 type resourceMeta struct {
+	resourceID   string
 	resourceURL  string
 	metrics      string
 	aggregations []string
 	resource     AzureResource
 }
 
-func (c *Collector) extractMetrics(ch chan<- prometheus.Metric, rm resourceMeta, httpStatusCode int, metricValueData AzureMetricValueResponse) {
+func (c *Collector) extractMetrics(ch chan<- prometheus.Metric, rm resourceMeta, httpStatusCode int, metricValueData AzureMetricValueResponse, publishedResource *map[string]bool) {
```
Comment (brian-brazil) [inline main.go]: You don't need a pointer to the map.

File: main.go [hunk 12]
```diff
@@ -120,39 +134,98 @@ func (c *Collector) batchCollectResources(ch chan<- prometheus.Metric, resources
...
 // Collect - collect results from Azure Montior API and create Prometheus metrics.
 func (c *Collector) Collect(ch chan<- prometheus.Metric) {
-
 	if err := ac.refreshAccessToken(); err != nil {
 		log.Println(err)
 		ch <- prometheus.NewInvalidMetric(azureErrorDesc, err)
 		return
 	}
 
+	var publishedResource = map[string]bool{}
```
Comment (brian-brazil) [inline main.go]: Why do you need to track this?

File: main.go (hunk 11 above)
Comment (rolandvarga) [reply main.go]: changed

File: main.go (hunk 12 above)
Comment (rolandvarga) [reply main.go]: We keep a list of resources already published here. This means we won't create the `azure_resource_info` metric multiple times for the same resource.

Comment (brian-brazil) [reply main.go]: You never use it in this function, you can move it down the call stack. Plural would also read better.

Comment (rolandvarga) [reply main.go]: changed based on your comments

**Comment by brian-brazil** [comment]
Thanks!

**Comment by rolandvarga** [comment]
> Thanks!

Thank you @brian-brazil
